src/main/kotlin/uk/gov/justice/digital/hmpps/subjectaccessrequestworker/services/pdf/PdfService.kt

```diff
@@ -0,0 +1,443 @@
+package uk.gov.justice.digital.hmpps.subjectaccessrequestworker.services.pdf
+
+import com.itextpdf.html2pdf.HtmlConverter
+import com.itextpdf.io.font.constants.StandardFonts
+import com.itextpdf.kernel.events.PdfDocumentEvent.END_PAGE
+import com.itextpdf.kernel.font.PdfFontFactory
+import com.itextpdf.kernel.pdf.PdfDocument
+import com.itextpdf.kernel.pdf.PdfReader
+import com.itextpdf.kernel.pdf.PdfWriter
+import com.itextpdf.kernel.utils.PdfMerger
+import com.itextpdf.layout.Document
+import com.itextpdf.layout.element.AreaBreak
+import com.itextpdf.layout.element.IBlockElement
+import com.itextpdf.layout.element.IElement
+import com.itextpdf.layout.element.Paragraph
+import com.itextpdf.layout.element.Text
+import com.itextpdf.layout.properties.AreaBreakType
+import com.itextpdf.layout.properties.TextAlignment
+import com.microsoft.applicationinsights.TelemetryClient
+import io.micrometer.common.util.StringUtils
+import org.apache.commons.lang3.time.StopWatch
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.subjectaccessrequestworker.config.trackSarEvent
+import uk.gov.justice.digital.hmpps.subjectaccessrequestworker.models.CustomHeaderEventHandler
+import uk.gov.justice.digital.hmpps.subjectaccessrequestworker.models.DpsService
+import uk.gov.justice.digital.hmpps.subjectaccessrequestworker.services.TemplateRenderService
+import uk.gov.justice.digital.hmpps.subjectaccessrequestworker.services.YamlFormatter
+import uk.gov.justice.digital.hmpps.subjectaccessrequestworker.utils.HeadingHelper
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+import java.io.OutputStream
+import java.time.LocalDate
+import java.time.format.DateTimeFormatter
+import java.time.format.FormatStyle
+
+@Service
+class PdfService(
+  val templateRenderService: TemplateRenderService,
+  val yamlFormatter: YamlFormatter,
+  val telemetryClient: TelemetryClient,
+) {
+
+  companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+    private val reportDateFormat = DateTimeFormatter.ofLocalizedDate(FormatStyle.LONG)
+
+    private const val BULLET_POINT = "\u2022"
+    private const val NO_BREAK_SPACE = "\u00a0"
+    private const val TOP_MARGIN = 50F
+    private const val RIGHT_MARGIN = 35F
+    private const val BOTTOM_MARGIN = 70F
+    private const val LEFT_MARGIN = 35F
+    private const val DATA_HEADER_FONT_SIZE = 16f
+    private const val DATA_FONT_SIZE = 12f
+    private const val DATA_LINE_SPACING = 16f
+  }
+
+  /**
+   * Generate a Subject Access Request report PDF document from the parameters provided
+   */
+  fun generateSubjectAccessRequestPDF(params: PdfParameters): CreatePdfResult {
+    telemetryClient.trackSarEvent(
```

Comment:
could the telemetry events be extracted to a method then methods would be cleaner?

Reply:
the telemetry code does get really dirty - I dislike it intently but couldn't work out a way to make code reusable but concise - short of just pulling each instance out as a single use method
I think when all is done and the service is stable there needs to be a ticket to remove all unnecessary telemetry